gtsfm/densify/patchmatchnet_data.py

```diff
@@ -0,0 +1,276 @@
+            self._images[self._patchmatchnet_idx_to_camera_idx[0]].width,
+        )
+        # Calculate the cropped size of each image so that the height and width of the image can be divided by 8 evenly,
+        #   so that the image size can be matched after downsampling and then upsampling in PatchmatchNet
+        self._cropped_h, self._cropped_w = (self._h - self._h % 8, self._w - self._w % 8)
+
+        self._pairs, self._depth_ranges = self.select_src_views_depth_ranges()
+
+    def select_src_views_depth_ranges(self) -> Tuple[np.ndarray, np.ndarray]:
+        """Configure (ref_view, src_view) pairs and depth_ranges for each view from sfm_result.
+        Ref: Wang et al. https://github.com/FangjinhuaWang/PatchmatchNet/blob/main/colmap_input.py (line 360-410)
+
+        If there are N0 valid images and the patchmatchnet's number of views is num_views, the function does:
+            1. Calculate the scores between each pair of the N0 images. Then for every image as the reference image,
+            find (num_views - 1) images with highest scores as the source images.
+            2. The scores between each pair of the N0 images is the sum of piecewise Gaussian scores between vectors
+            from a common track's coordinates to each camera center, which is mentioned in "View Selection" paragraphs
+            in Yao's paper https://arxiv.org/abs/1804.02505.
+            3. For every image as the reference image, calculate the depth range.
+
+        Returns:
+            pairs: array of shape (num_images, num_views-1). Each row_id indicates the index of reference view
+                in self.keys, with (num_views-1) values indicating the indices of source views in self.keys
+            depth_ranges: array of shape (num_images, 2). Each row_id indicates the index of reference view
+                in self.keys, with 2 values indicating [min_depth, max_depth]
+        """
+        num_tracks = self._sfm_result.number_tracks()
+
+        # Initialize the pairwise scores between the same views as negative infinity
+        pair_scores = np.zeros((self._num_images_cal, self._num_images_cal))
+        np.fill_diagonal(pair_scores, -np.inf)
+
+        # Initialize empty lists to collect all possible depths for each view
+        depths = defaultdict(list)
+
+        for j in range(num_tracks):
+            track = self._sfm_result.get_track(j)
+            num_measurements = track.number_measurements()
+            measurements = [track.measurement(k) for k in range(num_measurements)]
+            w_x = track.point3()
+            for k1 in range(num_measurements):
+                i_a = measurements[k1][0]
+                # Check if i_a is a valid image with estimated camera pose, then get its id for patchmatchnet
+                if i_a in self._camera_idx_to_patchmatchnet_idx:
+                    i_pm_a = self._camera_idx_to_patchmatchnet_idx[i_a]
+                else:
+                    i_pm_a = -1
```

Comment:
Thanks for all the updates, Ren. we should be able to consolidate the 4 lines above into a single line:
```python
# defaults to -1
i_pm_a = self._camera_idx_to_patchmatchnet_idx.get(i_a, -1)
```